proof-of-file-existence/tutorial/02-pofe-chain.md

```diff
@@ -0,0 +1,157 @@
+
+## Modifying our application
+
+The first thing we want to do is get rid of all the `Set` actions. In reality, you could keep these, but we won't be using them in our application.
+
+Start by deleting the `poe/x/pofe/handlerMsgSetClaim.go` and `poe/x/pofe/handlerMsgSetClaim.go` files.
+
+Once this is done, you should see some errors pop up in the window that is running your applcation, saying that it can't find references to the `MsgSetClaim` struct and the `handlerMsgSetClaim` function.
+
+This can be fixed by removing functions or lines of code that implement these in the files.
+
+
+Next, we want to implement an interface that allows someone to hash a file and submit the hash to the blockchain, without directly uploading its contents. We'll be implementing this via the command line in `./x/pofe/client/cli/txClaim.go`:
+
+```go
+package cli
+
+import (
+	"bufio"
+	"crypto/sha256"
+	"encoding/hex"
+	"io/ioutil"
+
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
+	"github.com/lukitsbrian/poe/x/pofe/types"
+)
+
+func GetCmdCreateClaim(cdc *codec.Codec) *cobra.Command {
+	return &cobra.Command{
+		Use:   "create-claim [path]",
+		Short: "Creates a new claim",
+		Args:  cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// accept a filepath
+			hasher := sha256.New()
+			s, _ := ioutil.ReadFile(args[0])
+			hasher.Write(s)
+			argsProof := hex.EncodeToString(hasher.Sum(nil))
+
+			cliCtx := context.NewCLIContext().WithCodec(cdc)
+			inBuf := bufio.NewReader(cmd.InOrStdin())
+			txBldr := auth.NewTxBuilderFromCLI(inBuf).WithTxEncoder(utils.GetTxEncoder(cdc))
+			msg := types.NewMsgCreateClaim(cliCtx.GetFromAddress(), string(argsProof))
+			err := msg.ValidateBasic()
+			if err != nil {
+				return err
+			}
+			return utils.GenerateOrBroadcastMsgs(cliCtx, txBldr, []sdk.Msg{msg})
+		},
+	}
+}
+```
+
+Lastly, instead of using the auto-generated uuid `ID` as our key, we will be using the `Proof` value in our struct instead. This will make it a lot easier to query an occurrence of `Proof` in our database. We can start by modifying the `CreateClaim` method in our `./x/pofe/keeper/claim.go` file to use `claim.Proof` instead of `claim.ID`.
+
+```go
+func (k Keeper) CreateClaim(ctx sdk.Context, claim types.Claim) {
+	store := ctx.KVStore(k.storeKey)
+	key := []byte(types.ClaimPrefix + claim.Proof)
+	value := k.cdc.MustMarshalBinaryLengthPrefixed(claim)
+	store.Set(key, value)
+}
+```
+
+According our application design, we would only need the `Creator` and the hash of the file. In this case, we can find and replace all instances of `ID` in our files. The easiest way to do this is via a text editor and search for all instances where `ID` is being used and remove them.
```

Comment:
Again, `ID` is not really a problem in this application, and removing it doesn't really add anything.

Reply:
To clarify on this a bit more - when running through this I replaced instances of `claim.ID` with `claim.Proof` where `ID` is still being used as a key (eg. keeper.Create), and removed `ID` from struct fields/method args where it is being used in addition `Proof` in order to make it compatible with the methods/types being used.

Alternatively I could keep ID in the struct (though it doesn't do anything), or use it as the key, but that might affect searchability, or have a section specifically for cleaning up the code.

What do you think?

Reply:
I think we do need to use hash (proof) as key, but we can keep the ID inside the struct to minimize the amount of work users have to do. I don't see a problem of having an extra field in the struct.